// inversify/lib/kernel/kernel.js
/**
 * inversify v.2.0.0-beta.1 - A lightweight IoC container written in TypeScript.
 * Copyright (c) 2015 Remo H. Jansen
 * MIT inversify.io/LICENSE
 * http://inversify.io
 */
"use strict";
var binding_count_1 = require("../bindings/binding_count");
var binding_1 = require("../bindings/binding");
var lookup_1 = require("./lookup");
var planner_1 = require("../planning/planner");
var resolver_1 = require("../resolution/resolver");
var ERROR_MSGS = require("../constants/error_msgs");
var METADATA_KEY = require("../constants/metadata_keys");
var binding_to_syntax_1 = require("../syntax/binding_to_syntax");
var metadata_1 = require("../planning/metadata");
var target_1 = require("../planning/target");
var request_1 = require("../planning/request");
var Kernel = (function () {
    function Kernel() {
        this._planner = new planner_1.default();
        this._resolver = new resolver_1.default();
        this._bindingDictionary = new lookup_1.default();
        this._middleware = null;
    }
    Kernel.prototype.load = function () {
        var _this = this;
        var modules = [];
        for (var _i = 0; _i < arguments.length; _i++) {
            modules[_i - 0] = arguments[_i];
        }
        modules.forEach(function (module) { module(_this); });
    };
    Kernel.prototype.applyMiddleware = function () {
        var middlewares = [];
        for (var _i = 0; _i < arguments.length; _i++) {
            middlewares[_i - 0] = arguments[_i];
        }
        this._middleware = middlewares.reverse().reduce(function (prev, curr) {
            return curr(prev);
        }, this._resolver.resolve.bind(this._resolver));
    };
    Kernel.prototype.bind = function (serviceIdentifier) {
        var binding = new binding_1.default(serviceIdentifier);
        this._bindingDictionary.add(serviceIdentifier, binding);
        return new binding_to_syntax_1.default(binding);
    };
    Kernel.prototype.unbind = function (serviceIdentifier) {
        try {
            this._bindingDictionary.remove(serviceIdentifier);
        }
        catch (e) {
            throw new Error(ERROR_MSGS.CANNOT_UNBIND + " " + serviceIdentifier);
        }
    };
    Kernel.prototype.unbindAll = function () {
        this._bindingDictionary = new lookup_1.default();
    };
    Kernel.prototype.get = function (serviceIdentifier) {
        return this._get(serviceIdentifier, null);
    };
    Kernel.prototype.getNamed = function (serviceIdentifier, named) {
        return this.getTagged(serviceIdentifier, METADATA_KEY.NAMED_TAG, named);
    };
    Kernel.prototype.getTagged = function (serviceIdentifier, key, value) {
        var metadata = new metadata_1.default(key, value);
        var target = new target_1.default(null, serviceIdentifier, metadata);
        return this._get(serviceIdentifier, target);
    };
    Kernel.prototype.getAll = function (serviceIdentifier) {
        var _this = this;
        var bindings = this._planner.getBindings(this, serviceIdentifier);
        switch (bindings.length) {
            case binding_count_1.default.NoBindingsAvailable:
                throw new Error(ERROR_MSGS.NOT_REGISTERED + " " + serviceIdentifier);
            case binding_count_1.default.OnlyOneBindingAvailable:
            case binding_count_1.default.MultipleBindingsAvailable:
            default:
                return bindings.map(function (binding) {
                    return _this._planAndResolve(binding, null);
                });
        }
    };
    Kernel.prototype.getServiceIdentifierAsString = function (serviceIdentifier) {
        var type = typeof serviceIdentifier;
        if (type === "function") {
            var _serviceIdentifier = serviceIdentifier;
            return _serviceIdentifier.name;
        }
        else if (type === "symbol") {
            return serviceIdentifier.toString();
        }
        else {
            var _serviceIdentifier = serviceIdentifier;
            return _serviceIdentifier;
        }
    };
    Kernel.prototype._get = function (serviceIdentifier, target) {
        var bindings = this._planner.getBindings(this, serviceIdentifier);
        if (target !== null) {
            var request = new request_1.default(serviceIdentifier, this._planner.createContext(this), null, bindings, target);
            bindings = this._planner.getActiveBindings(request, target);
        }
        if (bindings.length === binding_count_1.default.NoBindingsAvailable) {
            throw new Error(ERROR_MSGS.NOT_REGISTERED + " " + serviceIdentifier);
        }
        else if (bindings.length === binding_count_1.default.OnlyOneBindingAvailable) {
            return this._planAndResolve(bindings[0], target);
        }
        else {
            throw new Error(ERROR_MSGS.AMBIGUOUS_MATCH + " " + serviceIdentifier);
        }
    };
    Kernel.prototype._planAndResolve = function (binding, target) {
        var context = this._planner.createContext(this);
        this._planner.createPlan(context, binding, target);
        return (this._middleware !== null) ? this._middleware(context) : this._resolver.resolve(context);
    };
    return Kernel;
}());
Object.defineProperty(exports, "__esModule", { value: true });
exports.default = Kernel;
