// inversify/lib/syntax/binding_to_syntax.js
/**
 * inversify v.2.0.0-beta.1 - A lightweight IoC container written in TypeScript.
 * Copyright (c) 2015 Remo H. Jansen
 * MIT inversify.io/LICENSE
 * http://inversify.io
 */
"use strict";
var binding_in_when_on_syntax_1 = require("./binding_in_when_on_syntax");
var binding_when_on_syntax_1 = require("./binding_when_on_syntax");
var binding_type_1 = require("../bindings/binding_type");
var BindingToSyntax = (function () {
    function BindingToSyntax(binding) {
        this._binding = binding;
    }
    BindingToSyntax.prototype.to = function (constructor) {
        this._binding.type = binding_type_1.default.Instance;
        this._binding.implementationType = constructor;
        return new binding_in_when_on_syntax_1.default(this._binding);
    };
    BindingToSyntax.prototype.toValue = function (value) {
        this._binding.type = binding_type_1.default.Value;
        this._binding.cache = value;
        this._binding.implementationType = null;
        return new binding_when_on_syntax_1.default(this._binding);
    };
    BindingToSyntax.prototype.toConstructor = function (constructor) {
        this._binding.type = binding_type_1.default.Constructor;
        this._binding.implementationType = constructor;
        return new binding_when_on_syntax_1.default(this._binding);
    };
    BindingToSyntax.prototype.toFactory = function (factory) {
        this._binding.type = binding_type_1.default.Factory;
        this._binding.factory = factory;
        return new binding_when_on_syntax_1.default(this._binding);
    };
    BindingToSyntax.prototype.toAutoFactory = function (serviceIdentifier) {
        this._binding.type = binding_type_1.default.Factory;
        this._binding.factory = function (context) {
            return function () {
                return context.kernel.get(serviceIdentifier);
            };
        };
        return new binding_when_on_syntax_1.default(this._binding);
    };
    BindingToSyntax.prototype.toProvider = function (provider) {
        this._binding.type = binding_type_1.default.Provider;
        this._binding.provider = provider;
        return new binding_when_on_syntax_1.default(this._binding);
    };
    return BindingToSyntax;
}());
Object.defineProperty(exports, "__esModule", { value: true });
exports.default = BindingToSyntax;
