// inversify/lib/inversify.js
/**
 * inversify v.2.0.0-beta.1 - A lightweight IoC container written in TypeScript.
 * Copyright (c) 2015 Remo H. Jansen
 * MIT inversify.io/LICENSE
 * http://inversify.io
 */
"use strict";
var kernel_1 = require("./kernel/kernel");
exports.Kernel = kernel_1.default;
var injectable_1 = require("./annotation/injectable");
exports.injectable = injectable_1.default;
var tagged_1 = require("./annotation/tagged");
exports.tagged = tagged_1.default;
var named_1 = require("./annotation/named");
exports.named = named_1.default;
var inject_1 = require("./annotation/inject");
exports.inject = inject_1.default;
var multi_inject_1 = require("./annotation/multi_inject");
exports.multiInject = multi_inject_1.default;
var param_name_1 = require("./annotation/param_name");
exports.paramName = param_name_1.default;
var decorator_utils_1 = require("./annotation/decorator_utils");
exports.decorate = decorator_utils_1.decorate;
var constraint_helpers_1 = require("./syntax/constraint_helpers");
exports.traverseAncerstors = constraint_helpers_1.traverseAncerstors;
exports.taggedConstraint = constraint_helpers_1.taggedConstraint;
exports.namedConstraint = constraint_helpers_1.namedConstraint;
exports.typeConstraint = constraint_helpers_1.typeConstraint;
